tests/common-tests.h

```diff
@@ -6364,6 +6384,375 @@ TEST_CASE_METHOD(common_tests, "Into XML vector with several fetches", "[core][x
+        {
+            soci::blob write_blob(sql);
+
+            static_assert(sizeof(dummy_data) >= 10, "Underlying assumption violated");
+            write_blob.write_from_start(dummy_data, 10);
+
+            const int first_id = 42;
+
+            // Write and retrieve blob from/into database
+            sql << "insert into soci_test (id, b) values(:id, :b)", soci::use(first_id), soci::use(write_blob);
+
+            // Append to write_blob - these changes must not reflect in the BLOB stored in the DB
+            write_blob.append("ay", 2);
+            CHECK(write_blob.get_len() == 12);
+            char buf[15];
+            std::size_t read_bytes = write_blob.read_from_start(buf, sizeof(buf));
+            CHECK(read_bytes == 12);
+            for (std::size_t i = 0; i < 10; ++i)
+            {
+                CHECK(buf[i] == dummy_data[i]);
+            }
+            CHECK(buf[10] == 'a');
+            CHECK(buf[11] == 'y');
+
+
+            soci::blob read_blob(sql);
+            sql << "select b from soci_test where id = :id", soci::use(first_id), soci::into(read_blob);
+            CHECK(sql.got_data());
+
+            CHECK(read_blob.get_len() == 10);
+
+            std::size_t bytes_read = read_blob.read_from_start(buf, sizeof(buf));
+            CHECK(bytes_read == read_blob.get_len());
+            CHECK(bytes_read == 10);
+            for (std::size_t i = 0; i < bytes_read; ++i)
+            {
+                CHECK(buf[i] == dummy_data[i]);
+            }
+
+            // Update original blob and insert new db-entry (must not change previous entry)
+            const int second_id = first_id + 1;
+            write_blob.trim(0);
+            static_assert(sizeof(dummy_data) >= 15 + 5, "Underlying assumption violated");
+            write_blob.write_from_start(dummy_data + 15, 5);
+
+            sql << "insert into soci_test (id, b) values (:id, :b)", soci::use(second_id), soci::use(write_blob);
+
+            // First, check that the original entry has not been changed
+            sql << "select b from soci_test where id = :id", soci::use(first_id), soci::into(read_blob);
+            CHECK(read_blob.get_len() == 10);
+
+            // Then check new entry can be read
+            sql << "select b from soci_test where id = :id", soci::use(second_id), soci::into(read_blob);
+
+            bytes_read = read_blob.read_from_start(buf, sizeof(buf));
+            CHECK(bytes_read == read_blob.get_len());
+            CHECK(bytes_read == 5);
+            for (std::size_t i = 0; i < bytes_read; ++i)
+            {
+                CHECK(buf[i] == dummy_data[i + 15]);
+            }
+        }
+        SECTION("Binary data")
+        {
+            const std::uint8_t binary_data[12] = {0, 1, 2, 3, 4, 5, 6, 7, 22, 255, 250 };
+
+            soci::blob write_blob(sql);
+
+            std::size_t bytes_written = write_blob.write_from_start(reinterpret_cast<const char *>(binary_data), sizeof(binary_data));
```

Comment:
Why would it do that? Using any `T *` where `sizeof(T) != 1` is not supported. What we should use is make this use `std::byte *`, but that type is only available since C++17.

Reply:
Err, because it takes pointer to any data, not a character string. Using `char*` for buffer data is a C-ism from the seventies, even C stdio functions such as `fwrite()` take `void*` since a very long time.

And of course using `void*` is supported, what you can't do is to use pointer arithmetic on such pointers, but we don't need it.

Reply:
BLOB == **binary** large object. Binary data is operated upon in units of bytes. Everything else just lures users to specify and interpret sizes in multiples of `sizeof(T)` instead of in bytes, imo.

Reply:
Sorry, we seem to be speaking about completely different things and I have to admit that I don't understand what do you mean at all.

Again, a proper API for working with binary data is `read(void*)` and `write(const void*)`, there is really not much to discuss about here and it has nothing to do with the data being in bytes (of course it is, what else could it be). Using `char*` means that the application code has to use casts when using buffers of any other type, which is inconvenient. And `char` is just wrong, as we're not working with characters.

Reply:
What I mean is that people could likely be mislead that
```cpp
MyObject objects[5];
blob.write_from_start(objects, 5);
```
would do the correct thing, which it won't. This code only compiles in the `void *` design, whereas it wouldn't compile in the current `char *`-based API.

In any case, changing to a `void *`-based API could be seen as a breaking change on a formal level. However, for all practical purposes old could would simply keep working as it did before.

> And char is just wrong, as we're not working with characters.

I agree, but that is the commonly accepted crutch to work with sequences of bytes pre C++17 that finally introduced `std::byte`.

Anyway, I'll change to a `void *` based API.